File: 0187/solution.ts
function findRepeatedDnaSequences(s: string): string[] {
  const res = new Set<string>();
  const uniques = new Set<string>();
  for (let i = 0; i <= s.length - 10; i++) {
    const curr = s.slice(i, i + 10);
    if (res.has(curr)) continue;
    if (uniques.has(curr)) {
      res.add(curr);
    } else {
      uniques.add(curr);
    }
  }
  return [...res];
}

File: 0187/solution.py
class Solution:
    def findRepeatedDnaSequences(self, s: str) -> list[str]:
        left, right = 0, 9
        seen, res = set(), set()
        while right < len(s):
            temp = s[left : right + 1]
            if temp in seen:
                res.add(temp)
            else:
                seen.add(temp)
            right += 1
            left += 1
        return res

    # time O(n)
    # space O(n)
